submariner-operator/broker-multi-tenancy.md

```diff
@@ -0,0 +1,321 @@
+compatibility. The change does not introduce new functionality (e.g.,
+isolation and protection) beyond configurable namespace names. This allows
+for multiple namespaces to exist concurrently, in a cooperative manner
+(i.e., discretionary access control by referencing the correct namespace).
+2. Support stricter isolation between namespaces. This step builds upon the
+previous and creates relevant service accounts, RBAC policies, etc., to ensure
+users can only access their designated Broker namespace (previous step assumed
+cooperative access).
+3. Separate Broker namespace creation and management to separate the Offering
+and ClusterSet Admin roles and privileges. This could be done via a Broker (full)
+ reconciliation loop (note: currently the broker is not active, but acts
+ as a storage namespace only). Changes should be made to minimize disruption to existing
+workflows (e.g., hide changes behind the existing `subctl` command line).
+
+The first and second steps are mostly complete by
+[submariner-operator PR 1682](https://github.com/submariner-io/submariner-operator/pull/1682).
+
+## Design Details
+
+### Overview of Proposed Changes
+
+- Add a new “central” controller/operator called Axon controller (may think
+of a better name.) to the Broker cluster. The controller runs in a new,
+singleton, Axon namespace. This controller is responsible for provisioning
+submariner tenants (i.e., a ClusterSet). The `subctl` CLI will be extended
+to support tenant management by adding the following commands (see details below):
+  - `subctl deploy axon-cp`
+  - `subctl create|delete|get|list tenant [tenant_id]`
+- Create the submariner CRDs when deploying the Axon operator, assuming that
+tenants (i.e., ClusterSet Admins) will not have permissions to create those.
+The `deploy broker` command will ensure the CRDs exist.
+- A submariner tenant is represented by a tenant object in the Axon namespace
+ and a dedicated tenant namespace. Each tenant has a string ID and the namespace
+ is some constant prefix + tenant ID.
+- A SA is created per tenant and a *partial* `broker-info.subm` file is generated
+ by `subctl` upon tenant creation, with credentials (client secret) to access
+ the tenant namespace only.
+- Tenant can be deleted by deleting the tenant object, and the Axon operator
+ will clean up its resources (i.e., remove namespace and all contained objects)
+- The `deploy broker` command of subctl will require some changes but will
+ also maintain backwards compatibility with the single broker operation mode:
+  - Add an optional `namespace` field to the `.subm` file, to identify the
+  tenant. If the field exists, it takes precedence over the `--namespace` CLI
+  argument. If neither is provided, uses the default broker namespace.
+  - Use the access credentials and the Broker URL that are already specified in
+  the `.subm` file to access the Broker, if they exist. Otherwise - fallback to
+  details from kubecontext.
```

Comment:
I don't believe we do that even now (fallback to kubecontext), why would we introduce this?